### PR TITLE
labeler: change key-range config to a list of ranges

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### What problem does this PR solve?
Then TiDB uses table partition, a table's range can have multiple key-ranges. If we want to set label rule for a table, we need to update many rules.

### What is changed and how it works?
Support new type `key-ranges` to allow set multiple ranges for a rule.

### Check List

Tests
- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: server/schedule/labeler/labeler.go [hunk 1]
```diff
@@ -96,46 +96,75 @@ func (l *RegionLabeler) adjustRule(rule *LabelRule) error {
...
+		rs := make([]*KeyRangeRule, 0, len(rules))
+		for _, r := range rules {
+			rr, err := l.adjustKeyRangeRule(r)
+			if err != nil {
+				return err
+			}
+			rs = append(rs, rr)
 		}
-		rule.Rule = &r
+		rule.Rule = rs
```
Comment (rleungx) [inline server/schedule/labeler/labeler.go:118]: How about using another name for `rule.Rule`? See https://github.com/pingcap/tidb/pull/27660#pullrequestreview-753499559

File: server/schedule/labeler/rules.go [hunk 2]
```diff
@@ -32,6 +32,8 @@ type LabelRule struct {
 const (
 	// KeyRange is the rule type that labels regions by key range.
 	KeyRange = "key-range"
+	// KeyRanges is the rule type that specifies a list of key ranges.
+	KeyRanges = "key-ranges"
```
Comment (JmPotato) [inline server/schedule/labeler/rules.go:36]: The name `key-ranges` is too similar to the `key-range`, what about using `key-range-list` instead?

File: server/schedule/labeler/labeler.go (hunk 1 above)
Comment (disksing) [reply server/schedule/labeler/labeler.go:118]: I'm not quite undertand. Would you like to provide a better name?

File: server/schedule/labeler/rules.go (hunk 2 above)
Comment (disksing) [reply server/schedule/labeler/rules.go:36]: @rleungx suggests to keep only one type. So I'm going to merge the 2 types to one (name `key-range`, but is a list of ranges)

File: server/schedule/labeler/labeler.go (hunk 1 above)
Comment (rleungx) [reply server/schedule/labeler/labeler.go:118]: I have no idea about it =。=

File: server/schedule/labeler/labeler.go [hunk 3]
```diff
@@ -96,16 +96,12 @@ func (l *RegionLabeler) adjustRule(rule *LabelRule) error {
...
 	case KeyRange:
-		r, err := l.adjustKeyRangeRule(rule.Rule)
-		if err != nil {
-			return err
-		}
-		rule.Rule = r
-		return nil
-	case KeyRanges:
-		rules, ok := rule.Rule.([]map[string]interface{})
+		rules, ok := rule.Rule.([]interface{})
```
Comment (rleungx) [inline server/schedule/labeler/labeler.go:99]: How about change it to `rule.Rule.([]interface{})`

Comment (disksing) [reply server/schedule/labeler/labeler.go:99]: but it is not required to be plural. This field contains the rule logic of the rule. Its content type depends on the rule-type. Maybe we can use some abstract word, like `Data` or `Detail`.

Comment (rleungx) [reply server/schedule/labeler/labeler.go:99]: make sence

Comment (rleungx) [reply server/schedule/labeler/labeler.go:99]: Will it be changed in this PR?

Comment (disksing) [reply server/schedule/labeler/labeler.go:99]: use `Data`?

Comment (rleungx) [reply server/schedule/labeler/labeler.go:99]: ok

File: server/schedule/labeler/rules.go [hunk 4]
```diff
@@ -26,7 +26,7 @@ type LabelRule struct {
 	Index    int           `json:"index"`
 	Labels   []RegionLabel `json:"labels"`
 	RuleType string        `json:"rule_type"`
-	Rule     interface{}   `json:"rule"`
+	Data     interface{}   `json:"Data"`
```
Comment (rleungx) [inline server/schedule/labeler/rules.go:29]:
```suggestion
	Data     interface{}   `json:"data"`
```

**Review by rleungx — APPROVED** [review]
The rest LGTM.

**Review by JmPotato — APPROVED** [review]
LGTM. BTW, so the constraint that we need to make sure the number of `key-range` is even on `ruel.Data` is guaranteed by TiDB?

**Comment by disksing** [comment]
> LGTM. BTW, so the constraint that we need to make sure the number of `key-range` is even on `ruel.Data` is guaranteed by TiDB?

There is no such constraint. I think you have misunderstood it. The `makeKeyRanges` function is only used in tests.

**Pull request merged**